crates/red_knot_python_semantic/src/semantic_index/use_def.rs

```diff
@@ -413,22 +441,33 @@ impl<'db> Iterator for ConstraintsIterator<'_, 'db> {
 impl std::iter::FusedIterator for ConstraintsIterator<'_, '_> {}
 
 pub(crate) struct DeclarationsIterator<'map, 'db> {
-    all_definitions: &'map IndexVec<ScopedDefinitionId, Definition<'db>>,
+    all_definitions: &'map IndexVec<ScopedDefinitionId, Option<Definition<'db>>>,
+    visibility_constraints: &'map VisibilityConstraints<'db>,
     inner: DeclarationIdIterator<'map>,
-    may_be_undeclared: bool,
 }
 
-impl DeclarationsIterator<'_, '_> {
-    pub(crate) fn may_be_undeclared(&self) -> bool {
-        self.may_be_undeclared
-    }
+pub(crate) struct DeclarationWithConstraint<'map, 'db> {
+    pub(crate) declaration: Option<Definition<'db>>,
+    pub(crate) visibility_constraints: &'map VisibilityConstraints<'db>,
```

Comment:
Same question here as above; why does each individual `DeclarationWithConstraint` need to carry a copy of this reference?